packages/host/app/components/directory.gts

```diff
@@ -0,0 +1,104 @@
+import Component from '@glimmer/component';
+import { service } from '@ember/service';
+import type RouterService from '@ember/routing/router-service';
+import type CardService from '../services/card-service';
+import { action } from '@ember/object';
+import { on } from '@ember/modifier';
+import { fn } from '@ember/helper';
+import { eq } from '@cardstack/boxel-ui/helpers/truth-helpers';
+import { RealmPaths } from '@cardstack/runtime-common';
+import { directory, Entry } from '../resources/directory';
+//@ts-ignore cached not available yet in definitely typed
+import { cached } from '@glimmer/tracking';
+
+interface Args {
+  Args: {
+    openDirs: string | undefined;
+    polling: 'off' | undefined;
+    url: string;
+    path: string | undefined;
+  }
+}
+
+export default class Directory extends Component<Args> {
+  <template>
+    {{#each this.listing.entries key="path" as |entry|}}
```

Comment:
After each poll, the `entries` is a list of new objects. The old entry for a directory is not `===` the new entry for the same directory. By default, that means `#each` will destroy the old component and create a new component. That also means the child directory's resource gets destroyed and created fresh, with no last-known contents, so it renders empty.

By setting key="path" we're telling glimmer it can treat two entries with the same path as equal, so the component is not destroyed.

An alternative solution here would be to change the implementation of the `directory` resource so it provides stable objects during polling. This would take more code but would potentially be more convenient for consumers of the entries, since they wouldn't need to remember to set `key`. I don't think that's necessary right now, if we end up using this directory resource more widely we can consider it.

Reply:
Wow. Thank you, this is all good now.